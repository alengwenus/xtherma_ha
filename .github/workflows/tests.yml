name: Tests

on:
  push:
    branches:
      - main
  pull_request:
  schedule:
    - cron: "0 0 * * *"

env:
  DEFAULT_PYTHON: "3.13"

jobs:
  pre-commit:
    runs-on: "ubuntu-latest"
    name: Pre-commit and pytest
    steps:
      - name: Check out the repository
        uses: actions/checkout@v5

      - name: Set up Python ${{ env.DEFAULT_PYTHON }}
        uses: actions/setup-python@v6
        with:
          python-version: ${{ env.DEFAULT_PYTHON }}
          cache: pip

      - name: Install Python requirements
        run: |
          pip install -r requirements.txt -r dev-requirements.txt

      - name: Run pre-commit on all files
        run: |
          pre-commit run --all-files --show-diff-on-failure --color=always

      - name: Run pytest
        run: |
          python -m pytest --timeout=10 tests

  hacs:
    runs-on: "ubuntu-latest"
    name: HACS
    if: github.repository == 'Xtherma/xtherma_ha'
    steps:
      - name: Check out the repository
        uses: "actions/checkout@v5"

      - name: HACS validation
        uses: "hacs/action@22.5.0"
        with:
          category: "integration"
          ignore: "topics"

  hassfest:
    runs-on: "ubuntu-latest"
    name: Hassfest
    steps:
      - name: Check out the repository
        uses: "actions/checkout@v5"

      - name: Hassfest validation
        uses: "home-assistant/actions/hassfest@master"
