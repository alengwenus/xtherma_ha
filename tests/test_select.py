"""Tests for the Xtherma select platform."""

from unittest.mock import patch

import pytest
from homeassistant.components.select import (
    SelectEntity,
)
from homeassistant.const import Platform
from homeassistant.exceptions import HomeAssistantError
from pytest_homeassistant_custom_component.common import snapshot_platform

from custom_components.xtherma_fp.xtherma_client_common import XthermaReadOnlyError
from tests.helpers import (
    get_select_platform,
    provide_modbus_data,
    provide_rest_data,
)

from .conftest import init_integration, init_modbus_integration

SELECT_ENTITY_ID_002 = "select.test_entry_xtherma_config_operating_mode"
SELECT_ENTITY_ID_MODBUS_002 = "select.test_entry_xtherma_modbus_config_operating_mode"


@pytest.mark.parametrize("mock_rest_api_client", provide_rest_data(), indirect=True)
async def test_setup_select_rest_api(
    hass, entity_registry, snapshot, mock_rest_api_client
) -> None:
    """Test the setup of select platform using REST API."""
    with patch("custom_components.xtherma_fp._PLATFORMS", [Platform.SELECT]):
        entry = await init_integration(hass, mock_rest_api_client)

    await snapshot_platform(hass, entity_registry, snapshot, entry.entry_id)


@pytest.mark.parametrize("mock_modbus_tcp_client", provide_modbus_data(), indirect=True)
async def test_setup_select_modbus_tcp(
    hass, entity_registry, snapshot, mock_modbus_tcp_client
) -> None:
    """Test the setup of select platform using MODBUS TCP."""
    with patch("custom_components.xtherma_fp._PLATFORMS", [Platform.SELECT]):
        entry = await init_modbus_integration(hass, mock_modbus_tcp_client)

    await snapshot_platform(hass, entity_registry, snapshot, entry.entry_id)


@pytest.mark.parametrize("mock_rest_api_client", provide_rest_data(), indirect=True)
async def test_select_entity(hass, mock_rest_api_client):
    await init_integration(hass, mock_rest_api_client)
    platform = get_select_platform(hass)
    state = hass.states.get(SELECT_ENTITY_ID_002)
    assert state.state == "auto"
    entity = platform.entities.get(state.entity_id)
    assert entity is not None
    assert entity.icon == "mdi:brightness-auto"


@pytest.mark.parametrize("mock_rest_api_client", provide_rest_data(), indirect=True)
async def test_set_select_rest(hass, mock_rest_api_client):
    await init_integration(hass, mock_rest_api_client)
    platform = get_select_platform(hass)
    state = hass.states.get(SELECT_ENTITY_ID_002)
    assert state.state == "auto"
    entity = platform.entities.get(state.entity_id)
    assert entity is not None
    assert isinstance(entity, SelectEntity)
    # init_integration uses REST-API, which cannot write
    with pytest.raises(HomeAssistantError) as exc_info:
        await entity.async_select_option(entity.options[0])
    assert isinstance(exc_info.value.__cause__, XthermaReadOnlyError)


@pytest.mark.parametrize(
    "mock_modbus_tcp_client",  # This refers to the fixture
    provide_modbus_data(),
    indirect=True,  # This tells pytest to pass the parameter to the fixture
)
async def test_set_select_modbus(hass, mock_modbus_tcp_client):
    await init_modbus_integration(hass, mock_modbus_tcp_client)
    platform = get_select_platform(hass)
    state = hass.states.get(SELECT_ENTITY_ID_MODBUS_002)
    entity = platform.entities.get(state.entity_id)
    assert entity is not None
    assert isinstance(entity, SelectEntity)
    await entity.async_select_option(entity.options[0])
    kwargs = mock_modbus_tcp_client.write_register.call_args.kwargs
    # verify arguments passed to write_register()
    assert kwargs["address"] == 1
    assert kwargs["value"] == 0
    assert kwargs["slave"] == 1
