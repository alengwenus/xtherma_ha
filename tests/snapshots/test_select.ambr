# serializer version: 1
# name: test_setup_select_modbus_tcp[mock_modbus_tcp_client0][select.test_entry_xtherma_modbus_config_operating_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        'standby',
        'heating',
        'cooling',
        'water',
        'auto',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': None,
    'entity_id': 'select.test_entry_xtherma_modbus_config_operating_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:cogs',
    'original_name': 'Operating mode',
    'platform': 'xtherma_fp',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': '002',
    'unique_id': 'test_entry_xtherma-002',
    'unit_of_measurement': None,
  })
# ---
# name: test_setup_select_modbus_tcp[mock_modbus_tcp_client0][select.test_entry_xtherma_modbus_config_operating_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test_entry_xtherma_modbus_config Operating mode',
      'icon': 'mdi:brightness-auto',
      'options': list([
        'standby',
        'heating',
        'cooling',
        'water',
        'auto',
      ]),
      'parameter': '002',
    }),
    'context': <ANY>,
    'entity_id': 'select.test_entry_xtherma_modbus_config_operating_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'auto',
  })
# ---
# name: test_setup_select_modbus_tcp[mock_modbus_tcp_client0][select.test_entry_xtherma_modbus_config_sg_ready_request-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        'off',
        'normal',
        'block',
        'raise',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': None,
    'entity_id': 'select.test_entry_xtherma_modbus_config_sg_ready_request',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:cogs',
    'original_name': 'SG-Ready request',
    'platform': 'xtherma_fp',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': '808',
    'unique_id': 'test_entry_xtherma-808',
    'unit_of_measurement': None,
  })
# ---
# name: test_setup_select_modbus_tcp[mock_modbus_tcp_client0][select.test_entry_xtherma_modbus_config_sg_ready_request-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test_entry_xtherma_modbus_config SG-Ready request',
      'icon': 'mdi:cancel',
      'options': list([
        'off',
        'normal',
        'block',
        'raise',
      ]),
      'parameter': '808',
    }),
    'context': <ANY>,
    'entity_id': 'select.test_entry_xtherma_modbus_config_sg_ready_request',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_setup_select_rest_api[mock_rest_api_client0][select.test_entry_xtherma_config_operating_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        'standby',
        'heating',
        'cooling',
        'water',
        'auto',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': None,
    'entity_id': 'select.test_entry_xtherma_config_operating_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:cogs',
    'original_name': 'Operating mode',
    'platform': 'xtherma_fp',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': '002',
    'unique_id': 'test_entry_xtherma-002',
    'unit_of_measurement': None,
  })
# ---
# name: test_setup_select_rest_api[mock_rest_api_client0][select.test_entry_xtherma_config_operating_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test_entry_xtherma_config Operating mode',
      'icon': 'mdi:brightness-auto',
      'options': list([
        'standby',
        'heating',
        'cooling',
        'water',
        'auto',
      ]),
      'parameter': '002',
    }),
    'context': <ANY>,
    'entity_id': 'select.test_entry_xtherma_config_operating_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'auto',
  })
# ---
# name: test_setup_select_rest_api[mock_rest_api_client0][select.test_entry_xtherma_config_sg_ready_request-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        'off',
        'normal',
        'block',
        'raise',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': None,
    'entity_id': 'select.test_entry_xtherma_config_sg_ready_request',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:cogs',
    'original_name': 'SG-Ready request',
    'platform': 'xtherma_fp',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': '808',
    'unique_id': 'test_entry_xtherma-808',
    'unit_of_measurement': None,
  })
# ---
# name: test_setup_select_rest_api[mock_rest_api_client0][select.test_entry_xtherma_config_sg_ready_request-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test_entry_xtherma_config SG-Ready request',
      'icon': 'mdi:cancel',
      'options': list([
        'off',
        'normal',
        'block',
        'raise',
      ]),
      'parameter': '808',
    }),
    'context': <ANY>,
    'entity_id': 'select.test_entry_xtherma_config_sg_ready_request',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
